modules/graceful/manager.go

```diff
@@ -141,11 +151,14 @@ func (g *gracefulManager) RunAtHammer(ctx context.Context, hammer func()) {
 		}
 	}()
 }
-func (g *gracefulManager) doShutdown() {
+func (g *Manager) doShutdown() {
 	if !g.setStateTransition(stateRunning, stateShuttingDown) {
 		return
 	}
 	g.lock.Lock()
+	if g.shutdown == nil {
+		g.shutdown = make(chan struct{})
```

Comment:
Who's waiting for this `chan`? Don't `g.shutdown` need to be populated at creation? I see you're doing the same a few lines below for `g.done`, but I'm scratching my head. 🤔

Reply:
Loads of places are waiting for `g.shutdown`.  They all call `IsShutdown()`. In terms of delayed creation - I've followed the pattern in context.Context which seems always do delayed channel creation as you don't want to leave an open unused channel - although in this case as something will definitely be listening on the channel that may not be worth it.

`g.done` is similar.

Reply:
The fact that it was `nil` confused me. Seems convoluted to check for `nil` each time it's checked for closed. If it were created along with the manager, wouldn't `IsShutdown()` (which is repeatedly called by many tasks) be reduced to just returning `g.shutdown`?

Reply:
Yes but I was following the practices of context.Context which suggested not initialising the channel until it was needed. Initialising at the start should be fine though as we will always need them.